File: androidp1/app/src/main/java/com/preappointment1/app/data/AuthHelper.kt
package com.preappointment1.app.data

import android.util.Log
import com.preappointment1.app.data.api.ApiClient
import com.preappointment1.app.data.model.AuthRequest

object AuthHelper {
    private const val TAG = "LPM_AUTH"
    private const val DEVICE_PASSWORD = "secret_device_password"

    suspend fun ensureAuthenticated(): Boolean {
        if (SessionManager.getToken() != null) return true
        return refreshToken()
    }

    /** Clears stale token and obtains a fresh JWT (login, then register fallback). */
    suspend fun refreshToken(): Boolean {
        SessionManager.clearToken()
        val email = "${SessionManager.getOrCreateDeviceId()}@local.device"
        val request = AuthRequest(email = email, password = DEVICE_PASSWORD)

        return try {
            val response = ApiClient.authApiService.login(request)
            SessionManager.saveToken(response.token)
            Log.d(TAG, "Token refreshed (login)")
            true
        } catch (e: Exception) {
            try {
                val response = ApiClient.authApiService.register(request)
                SessionManager.saveToken(response.token)
                Log.d(TAG, "Token refreshed (register)")
                true
            } catch (registerError: Exception) {
                Log.e(TAG, "Token refresh failed", registerError)
                false
            }
        }
    }

    suspend fun ensureProfile(): String? {
        SessionManager.getProfileId()?.let { return it }

        return try {
            val profile = ApiClient.apiService.createProfile(
                com.preappointment1.app.data.model.ProfileRequest(
                    first_name = "Patient",
                    last_name = "Local",
                    relation = "Self"
                )
            )
            SessionManager.saveProfileId(profile.id)
            profile.id
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create profile", e)
            null
        }
    }
}

File: iosp1/P1/P1/Network/AuthHelper.swift
import Foundation

class AuthHelper {
    static let shared = AuthHelper()

    private init() {}

    func ensureAuthenticated() async -> Bool {
        if SessionManager.shared.getAccessToken() != nil {
            return true
        }
        return await refreshToken()
    }

    func refreshToken() async -> Bool {
        if let refresh = SessionManager.shared.getRefreshToken() {
            do {
                let response = try await ApiService.shared.refreshTokens(
                    request: RefreshTokenRequest(refreshToken: refresh),
                    allowAuthRetry: false
                )
                SessionManager.shared.saveTokens(access: response.accessToken, refresh: response.refreshToken)
                print("LPM_AUTH: Token refreshed (refresh token)")
                return true
            } catch {
                print("LPM_AUTH: Refresh token invalid, device auth fallback")
            }
        }

        SessionManager.shared.clearToken()
        return await DeviceAuth.signIn()
    }

    func ensureProfile() async -> String? {
        if let profileId = SessionManager.shared.getProfileId() {
            return profileId
        }

        do {
            let profile = try await ApiService.shared.createProfile(
                request: ProfileRequest(first_name: "Patient", last_name: "Local", relation: "Self")
            )
            SessionManager.shared.saveProfileId(profile.id)
            return profile.id
        } catch {
            print("LPM_AUTH: Failed to create profile - \(error)")
            return nil
        }
    }
}
